// contentful-management/dist/es-modules/methods/bulk-action.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _wrapNativeSuper(Class) { var _cache = typeof Map === "function" ? new Map() : undefined; _wrapNativeSuper = function _wrapNativeSuper(Class) { if (Class === null || !_isNativeFunction(Class)) return Class; if (typeof Class !== "function") { throw new TypeError("Super expression must either be null or a function"); } if (typeof _cache !== "undefined") { if (_cache.has(Class)) return _cache.get(Class); _cache.set(Class, Wrapper); } function Wrapper() { return _construct(Class, arguments, _getPrototypeOf(this).constructor); } Wrapper.prototype = Object.create(Class.prototype, { constructor: { value: Wrapper, enumerable: false, writable: true, configurable: true } }); return _setPrototypeOf(Wrapper, Class); }; return _wrapNativeSuper(Class); }

function _construct(Parent, args, Class) { if (_isNativeReflectConstruct()) { _construct = Reflect.construct; } else { _construct = function _construct(Parent, args, Class) { var a = [null]; a.push.apply(a, args); var Constructor = Function.bind.apply(Parent, a); var instance = new Constructor(); if (Class) _setPrototypeOf(instance, Class.prototype); return instance; }; } return _construct.apply(null, arguments); }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _isNativeFunction(fn) { return Function.toString.call(fn).indexOf("[native code]") !== -1; }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import { BulkActionStatus } from '../entities/bulk-action';
import { sleep } from './utils';
var DEFAULT_MAX_RETRIES = 30;
var DEFAULT_INITIAL_DELAY_MS = 1000;
var DEFAULT_RETRY_INTERVAL_MS = 2000;
export var BulkActionProcessingError = /*#__PURE__*/function (_Error) {
  _inherits(BulkActionProcessingError, _Error);

  var _super = _createSuper(BulkActionProcessingError);

  function BulkActionProcessingError(message, bulkAction) {
    var _this;

    _classCallCheck(this, BulkActionProcessingError);

    _this = _super.call(this, message);
    _this.bulkAction = bulkAction;
    _this.name = _this.constructor.name;
    return _this;
  }

  return BulkActionProcessingError;
}( /*#__PURE__*/_wrapNativeSuper(Error));
export var BulkActionFailedError = /*#__PURE__*/function (_BulkActionProcessing) {
  _inherits(BulkActionFailedError, _BulkActionProcessing);

  var _super2 = _createSuper(BulkActionFailedError);

  function BulkActionFailedError() {
    _classCallCheck(this, BulkActionFailedError);

    return _super2.apply(this, arguments);
  }

  return BulkActionFailedError;
}(BulkActionProcessingError);

/**
 * @private
 * @description Waits for a BulkAction to be completed and to be in one of the final states (failed or succeeded)
 * @param {Function} fn - GET function that will be called every interval to fetch a bulkAction status
 * @throws {BulkActionFailedError} throws an error if `throwOnFailedExecution = true` with the bulkAction that failed.
 * @throws {BulkActionProcessingError} throws an error with a bulkAction when processing takes too long.
 */
export function pollBulkActionStatus(_x, _x2) {
  return _pollBulkActionStatus.apply(this, arguments);
}

function _pollBulkActionStatus() {
  _pollBulkActionStatus = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(getBulkActionFunction, options) {
    var _options$retryCount, _options$retryInterva, _options$initialDelay, _options$throwOnFaile;

    var retryCount, done, action, maxRetries, retryIntervalMs, initialDelayMs, throwOnFailedExecution;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            retryCount = 0;
            done = false;
            maxRetries = (_options$retryCount = options === null || options === void 0 ? void 0 : options.retryCount) !== null && _options$retryCount !== void 0 ? _options$retryCount : DEFAULT_MAX_RETRIES;
            retryIntervalMs = (_options$retryInterva = options === null || options === void 0 ? void 0 : options.retryIntervalMs) !== null && _options$retryInterva !== void 0 ? _options$retryInterva : DEFAULT_RETRY_INTERVAL_MS;
            initialDelayMs = (_options$initialDelay = options === null || options === void 0 ? void 0 : options.initialDelayMs) !== null && _options$initialDelay !== void 0 ? _options$initialDelay : DEFAULT_INITIAL_DELAY_MS;
            throwOnFailedExecution = (_options$throwOnFaile = options === null || options === void 0 ? void 0 : options.throwOnFailedExecution) !== null && _options$throwOnFaile !== void 0 ? _options$throwOnFaile : true; // Initial delay for short-running BulkActions

            _context.next = 8;
            return sleep(initialDelayMs);

          case 8:
            if (!(retryCount < maxRetries && !done)) {
              _context.next = 22;
              break;
            }

            _context.next = 11;
            return getBulkActionFunction();

          case 11:
            action = _context.sent;

            if (!(action && [BulkActionStatus.succeeded, BulkActionStatus.failed].includes(action.sys.status))) {
              _context.next = 17;
              break;
            }

            done = true;

            if (!(action.sys.status === BulkActionStatus.failed && throwOnFailedExecution)) {
              _context.next = 16;
              break;
            }

            throw new BulkActionFailedError('BulkAction failed to execute.', action);

          case 16:
            return _context.abrupt("return", action);

          case 17:
            _context.next = 19;
            return sleep(retryIntervalMs);

          case 19:
            retryCount += 1;
            _context.next = 8;
            break;

          case 22:
            throw new BulkActionProcessingError("BulkAction didn't finish processing within the expected timeframe.", action);

          case 23:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _pollBulkActionStatus.apply(this, arguments);
}

/** Waits for a BulkAction status to be either succeeded or failed.
 * Used by the Plain client */
export function waitForBulkActionProcessing(_x3, _x4) {
  return _waitForBulkActionProcessing.apply(this, arguments);
}

function _waitForBulkActionProcessing() {
  _waitForBulkActionProcessing = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(_ref, options) {
    var plainClient, spaceId, environmentId, bulkActionId;
    return regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            plainClient = _ref.plainClient, spaceId = _ref.spaceId, environmentId = _ref.environmentId, bulkActionId = _ref.bulkActionId;
            return _context3.abrupt("return", pollBulkActionStatus( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
              return regeneratorRuntime.wrap(function _callee2$(_context2) {
                while (1) {
                  switch (_context2.prev = _context2.next) {
                    case 0:
                      return _context2.abrupt("return", plainClient.bulkAction.get({
                        bulkActionId: bulkActionId,
                        spaceId: spaceId,
                        environmentId: environmentId
                      }));

                    case 1:
                    case "end":
                      return _context2.stop();
                  }
                }
              }, _callee2);
            })), options));

          case 2:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _waitForBulkActionProcessing.apply(this, arguments);
}
